asteroid/utils/torch_utils.py

```diff
@@ -123,3 +123,107 @@ def are_models_equal(model1, model2):
         if p1.data.ne(p2.data).sum() > 0:
             return False
     return True
+
+
+def slice(sources, w_size=16384, h_size=16384 // 2):
+    """  Slice a signal in slices of size w_size
+
+     Args:
+         sources (torch.Tensor): sources of shape [nb_sources,channels,samples]
+         w_size (int): size of the window
+         h_size (int): size of the stride
+     Returns:
+         sliced signal (torch.Tensor): shape [nb_sources,nb_slices,channels,
+                                              samples]
+     """
+    # stride must be shorter than window
+    if h_size < w_size / 2:
+        raise ValueError(
+            f"You can't have a stride smaller than half the window")
+    # stride must be shorter than window
+    if h_size > w_size:
+        raise ValueError(
+            f"You can't have a stride larger than the window")
+    # Get source length
+    len_s = sources.size(-1)
+    # If source is longer than w_size slice
+    if len_s > w_size:
+        padding = 0
+        # Automatic padding
+        if (len_s - w_size) % h_size != 0:
+            padded = torch.zeros((sources.size(0), sources.size(1),
+                                  sources.size(2) + w_size - (len_s - w_size) %
+                                  h_size))
+            padded[:, :, :len_s] = sources
+            sources = padded
+            len_s = sources.size(-1)
+        nb_slices = int((len_s - w_size) // h_size) + 1
+        sliced = torch.zeros((sources.size(0),  nb_slices, sources.size(1),
+                              w_size))
+        for j in range(nb_slices):
+            sliced[:, j, :, :] = sources[:, :, j * h_size: j * h_size + w_size]
+        return sliced
+    # If source is smaller then pad it to w_size
+    if len_s < w_size:
+        new_size = list(sources.size()[:-1]) + [w_size]
+        padded = torch.zeros(new_size)
+        padded[:, :, :len_s] = sources
+        return padded.unsqueeze(1)
+    # Else source equal window just unsqueeze
+    return sources.unsqueeze(1)
+
+
+def unslice(sources, original_shape, w_size=16384, h_size=16384 // 2):
```

Comment:
Because we pad in the slicing, we should have a target_len argument to remove the zeros we added at the end right?

Reply:
Yes, that's what original_shape arg in unslice is. However, we can't remove the padding in the output of `slice() `otherwise the slicing method is unsuitable for models that only take a specific shape as input.

Reply:
Oh yes, sorry. 
Then this should be a kwarg with None as default. 
If None, we don't remove any padding.